Fix Goodwe spinlock

starkillerOG opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
As explained in this issue: https://github.com/home-assistant/core/issues/80145 the Goodwe integration is causing high CPU load between midnight and midnight + the UTC offset of the timezone.
I myself also see high CPU between midnight (00:00) and 2AM (02:00), I am in timezone UTC+2.

The next_midnight is not calculated correclty and gives the already pasted midnight (between midnigth and the UTC offset), this causes the callback to fire imediatly and this causes a spin of the callback which locks up the CPU.
~At least I highly suspect that is what is going on.~

I think the problem is utcnow is used instead of now, (utc vs local time).

~Still needs testing to make sure it is now fixed.~

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes https://github.com/home-assistant/core/issues/80145 https://github.com/home-assistant/core/issues/80588
- This PR is related to issue: https://github.com/home-assistant/core/issues/80216 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

starkillerOG committed:
fix spinlock

starkillerOG committed:
Add debug log

lyricnz commented:
SyntaxError: from __future__ imports must occur at the beginning of the file

Other than that it _looks_ like it might be scheduling the reset at the right time (currently ~15h from now for me).

Why are you adding days=1 in async_added_to_hass() and days=1,minutes=1 in async_reset()?

lyricnz commented:
I've applied the patch to my local, fiddled the timezone to hit midnight in 24 minutes, and am waiting for that to happen (24 minutes).  Will post History and debug logs (I added some) in about 30 minutes.

lyricnz commented:
OK midnight struck.

I had added the following code (below your calculation of next_midnight):

```
        next_midnight = dt_util.start_of_local_day(dt_util.now() + timedelta(days=1, minutes=1))
        now1 = dt_util.utcnow()
        _LOGGER.info(f"XXX1 utcnow {now1} {now1.tzinfo}")
        now2 = dt_util.now()
        _LOGGER.info(f"XXX1 {now2} {now2.tzinfo}")
        _LOGGER.info(f"XXX1 next_midnight {next_midnight} {next_midnight.tzinfo}")
```

And this output:

```
2022-10-21 08:00:05.688 INFO (MainThread) [homeassistant.components.goodwe.sensor] XXX1 utcnow 2022-10-20 21:00:05.688802+00:00 UTC
2022-10-21 08:00:05.689 INFO (MainThread) [homeassistant.components.goodwe.sensor] XXX1 2022-10-21 00:00:05.688977+03:00 Asia/Baghdad
2022-10-21 08:00:05.689 INFO (MainThread) [homeassistant.components.goodwe.sensor] XXX1 next_midnight 2022-10-22 00:00:00+03:00 Asia/Baghdad
```

Which looks right.  However it does't seem to have reset the "today's pv generation" value (which was at about 11.0kWh left over from yesterday)

![image](https://user-images.githubusercontent.com/122371/197057867-bd3e7b82-a8a3-4aa7-af6e-5c35040a81bb.png)

lyricnz commented:
I modified the timezone again to +2, and restarted HA with the following code-changes in place (note the two places of logging)

![image](https://user-images.githubusercontent.com/122371/197059789-595e83de-35d6-4833-aaea-018de09e86ed.png)

On startup it emitted:

```
2022-10-21 08:16:49.689 INFO (MainThread) [homeassistant.components.goodwe.sensor] XXX2 utcnow 2022-10-20 21:16:49.689742+00:00 UTC
2022-10-21 08:16:49.689 INFO (MainThread) [homeassistant.components.goodwe.sensor] XXX2 2022-10-20 23:16:49.689857+02:00 Europe/Kaliningrad
2022-10-21 08:16:49.689 INFO (MainThread) [homeassistant.components.goodwe.sensor] XXX2 next_midnight 2022-10-21 00:00:00+02:00 Europe/Kaliningrad
```

Will see what happens about 40 minutes (especially the XXX1 logging).

Currently Today's PV Generation is continuing to rise slowly as the sun comes up here.

lyricnz commented:
At the next hour:
```
2022-10-21 09:00:00.001 INFO (MainThread) [homeassistant.components.goodwe.sensor] XXX1 utcnow 2022-10-20 22:00:00.001618+00:00 UTC
2022-10-21 09:00:00.001 INFO (MainThread) [homeassistant.components.goodwe.sensor] XXX1 2022-10-21 00:00:00.001824+02:00 Europe/Kaliningrad
2022-10-21 09:00:00.001 INFO (MainThread) [homeassistant.components.goodwe.sensor] XXX1 next_midnight 2022-10-22 00:00:00+02:00 Europe/Kaliningrad
```

And the value remains un-reset

![image](https://user-images.githubusercontent.com/122371/197069622-3d090d94-ca84-4e16-8fdb-dd3dd2248166.png)

lyricnz commented:
Sorry I did not have level at debug, so could not  see if this was emitted
             _LOGGER.debug(f"Goodwe reset {self.name} to 0")

lyricnz commented:
OK starting once more, with timezone set to +1, which should fire in about 30 minutes.  Logging set to homeassistant.components.goodwe: debug

lyricnz commented:
Looks like the condition on the reset code above didn't match - there was no output from the message, just the rollover and the data didn't reset.

```
2022-10-21 10:00:00.000 INFO (MainThread) [homeassistant.components.goodwe.sensor] XXX1 utcnow 2022-10-20 23:00:00.000739+00:00 UTC
2022-10-21 10:00:00.001 INFO (MainThread) [homeassistant.components.goodwe.sensor] XXX1 2022-10-21 00:00:00.001128+01:00 Africa/Casablanca
2022-10-21 10:00:00.001 INFO (MainThread) [homeassistant.components.goodwe.sensor] XXX1 next_midnight 2022-10-22 00:00:00+01:00 Africa/Casablanca
```

lyricnz commented:
Why is intended to be the purpose of this code: `if not self.coordinator.last_update_success`?

It's perfectly reasonable for the last update to be success, isn't it?  For example, above the arctic circle, there may still be sunlight at midnight.  Or future versions of Goodwe may have mains-backup, and not go completely dead when there is no solar.

I hacked the code to apply midnight in +60 seconds (quicker to test), and to fail to fetch the data from goodwe, and the reset worked then. So suspect the check is incorrect/unnecessary?

lyricnz commented:
Hmm. When I reran the code, with just your patch, and the `if not self.coordinator.last_update_success` check removed, the data has  single zero-sample, then goes back to the value returned from the inverter.  I know this is an edge case, but will be something to watch out for.

![image](https://user-images.githubusercontent.com/122371/197078129-c00afad4-b829-455a-bc5f-bab87248c5c3.png)

starkillerOG committed:
fix styling

starkillerOG commented:
> SyntaxError: from **future** imports must occur at the beginning of the file
> 
> Other than that it _looks_ like it might be scheduling the reset at the right time (currently ~15h from now for me).
> 
> Why are you adding days=1 in async_added_to_hass() and days=1,minutes=1 in async_reset()?

Yea I did not do styiling yet, just fixed that.

Because async_added_to_hass() will happen at a random time during the day, and the async_reset will always occur at 00:00:00 so in order to prevent confusion I added a extra minute to have it at 00:01:00 and then have dt_util.start_of_local_day go back to the correct midnight time.

starkillerOG commented:
I tested this code and it seems to solve the issue with high CPU load.
From the debug log I can see the reset function now only executed once at midnight (with UTC+2 timezone).

starkillerOG commented:
@lyricnz bay the way, the reset will not happen when you reload the goodwe integration after the invertor went offline, since the integration will not be able to be set-up, the daily energy producition sensor will not be set up, and the reset callback will also not be scheduled.

starkillerOG commented:
This is ready to be merged.

starkillerOG commented:
@MartinHjelmare could you add this bugfix to the 2022.10.6 milestone?
